### PR TITLE
Arcs system tracing on DevEx

### PR DESCRIPTION
**Pull request by ianull**
[Arcs System Trace on DevEx](http://go/arcs-systrace-devex)

This implements a quick path to Arcs system tracing capabilities (roughly stage A & B) at [the design](http://go/arcs-systrace).

### CODE REVIEW

**Comment by ianull** [comment]
Hope one of you can help review post-holidays.
Happy New Year!

File: src/tracelib/systrace.ts [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+/**
+ * Used at sources that cannot decorate class by @SystemTrace.
+ * Extends SystemTraceable to declare your class system-traceable,
+ * e.g. class Foo extends SystemTraceable {...}
+ */
+// tslint:disable-next-line: variable-name
+export const SystemTraceable = SystemTrace(class {});
+
+/**
+ * Used at sources that cannot decorate class by @SystemTrace.
```
Comment (shaper) [inline src/tracelib/systrace.ts]: What is an example such source? I am just curious what catalyzed this.

File: javaharness/java/arcs/android/AndroidArcsEnvironment.java [hunk 2]
```diff
@@ -146,6 +146,7 @@ public WebResourceResponse shouldInterceptRequest(WebResourceRequest request) {
     });
 
     webView.addJavascriptInterface(this, "DeviceClient");
+    webView.addJavascriptInterface(new ArcsSystemTrace(), "systemTraceApis");
```
[the remark this answers is not in the document]
Comment (ianull) [reply javaharness/java/arcs/android/AndroidArcsEnvironment.java]: Hmm, it makes sense to maintain capitalized naming, I'll change it.
As for why not just SystemTrace, I thought about it. To differentiate it from the @SystemTrace decorator in JS, SystemTraceApis can tell it's apis imported from somewhere like Java world or other delegations instead of built-in functions in JS.

File: javaharness/java/arcs/android/AndroidRuntimeSettings.java [hunk 3]
```diff
@@ -44,6 +46,8 @@
   private static final int DEFAULT_DEV_SERVER_PORT = 8786;
   // Activates the Arcs Cache Manager.
   private static final boolean DEFAULT_USE_CACHE_MANAGER = true;
+  // Disables system trace
```
[the remark this answers is not in the document]
Comment (ianull) [reply javaharness/java/arcs/android/AndroidRuntimeSettings.java]: The details/definitions are put in RuntimeSettings.java as:
  // Available options:
  //   'android': trace messages are bridged to the Android Trace APIs
  //   'console': trace messages are bridged to JS console
  // Options not listed above would be skipped, namely trace messages are
  // neither generated nor bridged.

AndroidRuntimeSettings implements RuntimeSettings and provisions the default values.

File: javaharness/java/arcs/android/ArcsSystemTrace.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package arcs.android;
+
+import android.os.Trace;
+import android.webkit.JavascriptInterface;
+
+/**
+ * Bridging the android.os.Trace APIs to the Arcs runtime/shell.
```
[the remark this answers is not in the document]
Comment (ianull) [reply javaharness/java/arcs/android/ArcsSystemTrace.java]: np, will add the [text]{@link namepathOrURL}

File: javaharness/java/arcs/android/ArcsSystemTrace.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * TODO(ianchang):
+ * Event tag-cookie pairs can be extended to implement more valuable
+ * performance metrics i.e. latency stats (percentile, average and so forth).
+ * and also be integrated with the existing logging and profiling frameworks
+ * e.g., westworld, perfgate and so on.
```
[the remark this answers is not in the document]
Comment (ianull) [reply javaharness/java/arcs/android/ArcsSystemTrace.java]: ok!

File: javaharness/java/arcs/android/ArcsSystemTrace.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ */
+final class ArcsSystemTrace {
+
+  public ArcsSystemTrace() {}
```
[the remark this answers is not in the document]
Comment (ianull) [reply javaharness/java/arcs/android/ArcsSystemTrace.java]: Right, a nullary constructor is created automatically. Thanks!

File: src/runtime/api-channel.ts [hunk 7]
```diff
@@ -550,6 +556,9 @@ export abstract class PECOuterPort extends APIPort {
   // TODO(sjmiles): experimental `services` impl
   abstract onServiceRequest(particle: recipeParticle.Particle, request: {}, callback: number);
 
+  abstract onSystemTraceBegin(tag: string, cookie: number);
```
[the remark this answers is not in the document]
Comment (ianull) [reply src/runtime/api-channel.ts]: I got your points.
The reason of naming them begin/end is just to align the naming at Android system trace.
So we have the same naming rules among imported Android Apis and internal Apis.

File: src/tracelib/systrace-clients.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {getGlobalScope, getSystemTraceApis, getSystemTraceChannel} from './systrace-helpers.js';
+
+const CONSOLE_CLIENT_NAME = 'console';
```
[the remark this answers is not in the document]
Comment (ianull) [reply src/tracelib/systrace-clients.ts]: sure, np.

File: src/tracelib/systrace-clients.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+  asyncTraceEnd(tag: string, cookie: number) {
+    console.timeStamp(`F|${tag}`);
+  }
+}
+
+/**
+ * Client: Android Arcs Tracing
+ *
+ * This client implement more coarse performance/latency measurement
```
[the remark this answers is not in the document]
Comment (ianull) [reply src/tracelib/systrace-clients.ts]: More coarse (vs. fine-grained) implies more inaccurate timestamp.
Say a traced function gets traced in a dedicated worker thread.
The message is delivered from the worker to the main renderer who in turns parses the message and dispatches the message to corresponding Java Apis via proxying the messages to the JavaBridge thread at the browser process.
Then the JavaBridge thread (singleton) handles the messages and invokes actual Android tracing Apis. The current Android tracing Apis (via underlying ftrace file nodes) uses the timestamp when writing to the underlying the ftrace file nodes instead the original timestamp when calling the function at the worker.

File: src/tracelib/systrace.ts [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+type Constructor<T> = new (...args) => T;
+
+// Identifies whether a class has already been traced in its
+// inheritance hierarchies.
+const SYSTEM_TRACED_PROPERTY = '_systemTraced';
+
+// Specifies the functions of a prototype/constructor not being traced.
+const DONT_TRACE_PROPERTY = '_dontTrace';
```
[the remark this answers is not in the document]
Comment (ianull) [reply src/tracelib/systrace.ts]: sure, will rename all 'dont' to 'no', thanks.

File: src/tracelib/systrace.ts (hunk 1 above)
Comment (ianull) [reply src/tracelib/systrace.ts]: Hmm, 
in typescript, use @SystemTrace to decorate the traced classes.
in javascript, no decorators, so we have to do:
  class Foo extends SystemTrace(class {}) {...}
It's non-trivial to devs to declare their traced class like this, instead:
  class Foo extends SystemTraceable {...}

File: src/tracelib/systrace-clients.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (ianull) [reply src/tracelib/systrace-clients.ts]: I'd better rephrase it to vague instead of coarse :-p

**Comment by ianull** [comment]
Looks like only Walter and Maria are in town today and tomorrow.
Walter has been the 1st reviewer, @mmandlis can you be the 2nd reviewer if you have time for it :)
Thanks.

**Comment by ianull** [comment]
Thank you!
